pkg/vz/network_darwin.go

```diff
@@ -50,77 +51,117 @@ func DialQemu(unixSock string) (*os.File, error) {
-	if err != nil {
-		logrus.Errorln("Failed to read header", err)
+// Read reads a QEMU packet and returns the contained raw packet.  Returns (len,
+// nil) if a packet was read, and (0, err) on error. Errors means the prorocol
+// is broken and the socket must be closed.
+func (c *qemuPacketConn) Read(b []byte) (n int, err error) {
+	var size uint32
+	if err := binary.Read(c.Conn, binary.BigEndian, &size); err != nil {
+		// Likely connection closed by peer.
+		return 0, err
 	}
 
-	size := binary.BigEndian.Uint32(header)
-	reader := io.LimitReader(v.unixConn, int64(size))
+	reader := io.LimitReader(c.Conn, int64(size))
 	_, err = reader.Read(b)
 	if err != nil {
-		logrus.Errorln("Failed to read packet", err)
+		// Likely connection closed by peer.
+		return 0, err
 	}
 	return int(size), nil
 }
 
-// Write puts QEMU header packet first and then writes the raw packet.
-func (v *QEMUPacketConn) Write(b []byte) (n int, err error) {
-	header := make([]byte, 4)
-	binary.BigEndian.PutUint32(header, uint32(len(b)))
-	_, err = v.unixConn.Write(header)
-	if err != nil {
-		logrus.Errorln("Failed to write header", err)
+// Write writes a QEMU packet containing the raw packet. Returns (len(b), nil)
+// if a packet was written, and (0, err) if a packet was not fully written.
+// Errors means the prorocol is broken and the socket must be closed.
+func (c *qemuPacketConn) Write(b []byte) (int, error) {
+	size := len(b)
+	header := uint32(size)
+	if err := binary.Write(c.Conn, binary.BigEndian, header); err != nil {
+		return 0, err
 	}
 
-	write, err := v.unixConn.Write(b)
-	if err != nil {
-		logrus.Errorln("Failed to write packet", err)
+	start := 0
+	for start < size {
+		nw, err := c.Conn.Write(b[start:])
+		if err != nil {
+			return 0, err
+		}
+		start += nw
 	}
-	return write, nil
+	return size, nil
 }
 
-func (v *QEMUPacketConn) Close() error {
-	return v.unixConn.Close()
-}
-
-func (v *QEMUPacketConn) LocalAddr() net.Addr {
-	return v.unixConn.LocalAddr()
-}
-
-func (v *QEMUPacketConn) RemoteAddr() net.Addr {
-	return v.unixConn.RemoteAddr()
-}
-
-func (v *QEMUPacketConn) SetDeadline(t time.Time) error {
-	return v.unixConn.SetDeadline(t)
-}
+// Testing show that retries are very rare (e.g 24 or 208,330 packets) and
+// requires 1 or 2 retries to complete the write. A 100 microseconds sleep loop
+// consumes about 4% CPU on M1 Pro.
+const writeRetryDelay = 100 * time.Microsecond
 
-func (v *QEMUPacketConn) SetReadDeadline(t time.Time) error {
-	return v.unixConn.SetReadDeadline(t)
+// packetConn handles ENOBUFS errors when writing to unixgram socket.
+type packetConn struct {
+	net.Conn
 }
 
-func (v *QEMUPacketConn) SetWriteDeadline(t time.Time) error {
-	return v.unixConn.SetWriteDeadline(t)
+// Write write a packet retrying on ENOBUFS errors.
+func (c *packetConn) Write(b []byte) (int, error) {
+	var retries uint64
+	for {
+		n, err := c.Conn.Write(b)
+		if err != nil && errors.Is(err, syscall.ENOBUFS) {
```

Comment:
Need to handle the case where n > 0 and err is syscall.ENOBUFS

Reply:
This is not possible with SOCK_DGRAM socket. You either send a complete datagram, or cannot send anything.

Maybe retry on `n == 0 && errors.Is(err, syscall.ENOBUFS)`, and let the call fail otherwise?